File: src/main/java/abstract_factory_java/MotherInLaw.java
package abstract_factory_java;

import abstract_factory_java.factory.PropertyFactory;
import abstract_factory_java.factory.impl.ProgrammerPropertyFactory;
import abstract_factory_java.property.Car;
import abstract_factory_java.property.House;
import abstract_factory_java.property.Money;

/**
 * @author angus
 * @date 2019/11/21
 */
public class MotherInLaw {
    public static void main(String[] args) {
        // 具体工厂可以运行时指定
        PropertyFactory factory = new ProgrammerPropertyFactory();
        Car car = factory.buyCar();
        House house = factory.buyHouse();
        Money money = factory.drawMoney();
        // 打印资产
        System.out.println("car: " + car);
        System.out.println("house: " + house);
        System.out.println("money: " + money);
    }
}

File: src/main/kotlin/abstract_factory/MotherInLaw.kt
package abstract_factory

import abstract_factory.factory.PropertyFactory
import abstract_factory.factory.impl.CommonPropertyFactory
import abstract_factory.factory.impl.ProgrammerPropertyFactory

/**
 * @author angus
 * @date 2019/11/21
 */
class MotherInLaw

fun main() {
    var factory: PropertyFactory = CommonPropertyFactory()
    var car = factory.buyCar()
    var house = factory.buyHouse()
    var money = factory.drawMoney()
    println("car: $car, house: $house, money: $money")

    // 也可以传递一个枚举类型参数
    factory = PropertyFactory.instance<ProgrammerPropertyFactory>()
    car = factory.buyCar()
    house = factory.buyHouse()
    money = factory.drawMoney()
    println("car: $car, house: $house, money: $money")
}